lib/collection/set.js

```diff
@@ -41,31 +39,35 @@ export class SetColl extends Collection {
   }
 
   remove(item) {
-    this._removeWithoutObserver(item);
-    this._notifyRemoved([item]);
+    if (this._removeWithoutObserver(item)) {
+      this._notifyRemoved([item]);
+    }
   }
 
   _removeWithoutObserver(item) {
-    this._set.delete(item);
+    return this._set.delete(item);
   }
 
   addAll(items) {
-    for (let item of items) {
-      this._addWithoutObserver(item);
+    items = items.filter(item => this._addWithoutObserver(item));
+    if (items.length) {
+      this._notifyAdded(items);
     }
-    this._notifyAdded(items);
   }
 
   removeAll(items) {
-    for (let item of items) {
-      this._removeWithoutObserver(item);
+    items = items.filter(item => this._removeWithoutObserver(item));
+    if (items.length) {
+      this._notifyRemoved(items);
     }
-    this._notifyRemoved(items);
   }
 
   clear() {
-    this._notifyRemoved(this.contents);
-    this._set.clear();
+    if (this._set.size) {
+      let items = this.contents;
+      this._set.clear();
+      this._notifyRemoved(items);
+    }
```

Comment:
As above: Please use early exits for errors and edge cases.

```suggestion
    if (this.isEmpty) {
      return;
    }
    let items = this.contents;
    this._set.clear();
    this._notifyRemoved(items);
```